### ISSUE TITLE
Very slow download?

### ISSUE THREAD

**Issue by LeafarDev** [Hexer10/youtube_explode_dart#41]
**Bug:**

It looks like it's downloading each piece every 10 seconds

**Code:**
```
import 'dart:async';
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:path_provider/path_provider.dart';
import 'package:youtube_explode_dart/youtube_explode_dart.dart';

// Initialize the YoutubeExplode instance.
final yt = YoutubeExplode();

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  print('Type the video id or url: ');

  var url = "https://www.youtube.com/watch?v=I7pBaCorMfM";

  // Save the video to the download directory.
  var dir = await getExternalStorageDirectory();
  Directory(dir.path).createSync();

  // Download the video.
  await download(url);

  yt.close();
  exit(0);
}

Future<void> download(String id) async {
  // Get video metadata.
  var video = await yt.videos.get(id);

  // Get the video manifest.
  var manifest = await yt.videos.streamsClient.getManifest(id);
  var streams = manifest.audioOnly;

  // Get the last audio track (the one with the highest bitrate).
  var audio = streams.last;
  var audioStream = yt.videos.streamsClient.get(audio);

  // Compose the file name removing the unallowed characters in windows.
  var fileName = '${video.title}.${audio.container.name.toString()}'
      .replaceAll('Container.', '')
      .replaceAll(r'\', '')
      .replaceAll('/', '')
      .replaceAll('*', '')
      .replaceAll('?', '')
      .replaceAll('"', '')
      .replaceAll('<', '')
      .replaceAll('>', '')
      .replaceAll('|', '');
  var dir = await getExternalStorageDirectory();
  print(dir.path);
  var file = File('${dir.path}/$fileName');

  // Create the StreamedRequest to track the download status.

  // Open the file in appendMode.
  var output = file.openWrite(mode: FileMode.writeOnlyAppend);

  // Track the file download status.
  var len = audio.size.totalBytes;
  var count = 0;
  var oldProgress = -1;

  // Create the message and set the cursor position.
  var msg = 'Downloading `${video.title}`(.${audio.container.name}):  \n';
  print(msg);
  //  var row = console.cursorPosition.row;
//  var col = msg.length - 2;
//  console.cursorPosition = Coordinate(row, 0);
//  console.write(msg);

  // Listen for data received.
  await for (var data in audioStream) {
    count += data.length;
    var progress = ((count / len) * 100).round();
    if (progress != oldProgress) {
//      console.cursorPosition = Coordinate(row, col);
      print('$progress%');
      oldProgress = progress;
    }
    output.add(data);
  }
  await output.close();
}
```

**Flutter Doctor**

```[✓] Flutter (Channel stable, v1.17.3, on Linux, locale en_US.UTF-8)
    • Flutter version 1.17.3 at /home/rafael/flutter
    • Framework revision b041144f83 (2 weeks ago), 2020-06-04 09:26:11 -0700
    • Engine revision ee76268252
    • Dart version 2.8.4

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /home/rafael/Android/Sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /home/rafael/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6325121/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Android Studio (version 4.1)
    • Android Studio at /home/rafael/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6325121
    • Flutter plugin version 43.0.3
    • Dart plugin version 193.5731
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[!] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /home/rafael/.local/share/JetBrains/Toolbox/apps/IDEA-U/ch-0/193.6911.18
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (1 available)
    • Redmi Note 5 • 41e4e3c • android-arm64 • Android 9 (API 28)

! Doctor found issues in 1 category.
```

**Comment by Hexer10** [comment]
I could reproduce this, this is happening because  ratebypass parameters is not parsed correctly, gonna push a fix asap

**Comment by Hexer10** [comment]
Please check if it is faster with v`1.2.2`

**Comment by LeafarDev** [comment]
Much better now !

**Comment by LeafarDev** [comment]
Hi again. Sounds like the file is partially corrupted, can you reproduce same issue?

**Comment by Hexer10** [comment]
Mh, I couldn't find any major issues, is it randomly corrupted? Are missing some parts at the start, middle or end?

**Comment by LeafarDev** [comment]
It's happening with long videos like  [https://www.youtube.com/watch?v=lHZA_-kJigY&t=27s](url). When trying to seek in the middle of audio the player crash or go back to 00:00.

**Comment by Hexer10** [comment]
Seems like that I download the streams with the range headers every `9898989` bytes, one byte is duplicated, and I think this might cause some issues...

A quick side note, I forgot to update the docs but instead of `.last` to get the best audio stream you should use `.withHighestBitrate()`

**Comment by Hexer10** [comment]
Pushed v`1.2.3`, which should address this, please let me know if the issue still persists.
I couldn't test this since atm I have a very slow connection and it would take me years to download that video

**Comment by mymikemiller** [comment]
I'm still having very slow downloads on 1.3.1. For example, [this](https://www.youtube.com/watch?v=Nt6h2EKQE68) three minute video takes over a half hour to download (and often fails partway through and has to be restarted due to an HttpException, but that's [another issue](https://github.com/Hexer10/youtube_explode_dart/issues/44)). My download code is in the linked issue.

If I pause and look at `videoStreamInfo.url`, I get a url like [this](https://r1---sn-bvvbax4pcxg-naje.googlevideo.com/videoplayback?expire=1592956024&ei=GEDyXs7KKpWFkgbj_KrIBg&ip=2605%3Aa601%3Aa9b3%3Af900%3Ae43a%3Aa8f1%3A7e7e%3A1234&id=o-ALVyM_Dl3_KzmFDfotbDH73uQnOYzL4ItMMr-540cqzC&itag=18&source=youtube&requiressl=yes&mh=0K&mm=31%2C29&mn=sn-bvvbax4pcxg-naje%2Csn-vgqsknes&ms=au%2Crdu&mv=m&mvi=0&pl=42&initcwndbps=1311250&vprv=1&mime=video%2Fmp4&gir=yes&clen=17811863&ratebypass=yes&dur=217.361&lmt=1574798172863215&mt=1592934337&fvip=1&c=WEB&txp=5531432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cratebypass%2Cdur%2Clmt&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRgIhALiyCIUH5di8pyLo9dbqvFRJ5IpEz4FgHpZtWGWnoR4aAiEAqjTa-PaoEpWS6JMX1vBcLwxtaJH889TVJWAnZCc-dBM%3D&sig=AOq0QJ8wRgIhAM8RiYRmODhNCkY29hNZYYTdcWFlWduycWY1owCSMYHaAiEA87XiSqWqRPpGFvdu5yjnSr1jI-Fc4Xnt92QaVt457Zg%3D), and when I go to that url in the browser, the entire video downloads in seconds. 

Should I skip calling _yt.videos.streamsClient.get(videoStreamInfo) and just make a standard http request to download the video? Or maybe use a library like [dio](https://pub.dev/packages/dio); people seem to like that one. But then I'm only using youtube_explode to get to the links, not to do the actual download. Are there any potential issues doing it that way? Maybe I'm more likely to be blocked by YouTube if I download directly?

**Comment by mymikemiller** [comment]
Edit: downloading the above three minute video actually takes about 30 seconds. The first ~20% of the video is downloaded in a few seconds, and the remaining 80% takes about 25.
